meinberlin/apps/kiezradar/serializers.py

```diff
@@ -0,0 +1,111 @@
+from rest_framework import serializers
+
+from adhocracy4.administrative_districts.models import AdministrativeDistrict
+from adhocracy4.projects.models import Topic
+from meinberlin.apps.kiezradar.models import KiezradarQuery
+from meinberlin.apps.kiezradar.models import ProjectType
+from meinberlin.apps.kiezradar.models import SearchProfile
+from meinberlin.apps.organisations.models import Organisation
+
+
+class SearchProfileSerializer(serializers.ModelSerializer):
+    """Serializer for the SearchProfile model."""
+
+    query = (
+        serializers.PrimaryKeyRelatedField(
+            queryset=KiezradarQuery.objects.all(), required=False
+        ),
+    )
+    query_text = serializers.CharField(required=False)
+    organisations = serializers.PrimaryKeyRelatedField(
+        queryset=Organisation.objects.all(), many=True, allow_null=True, required=False
+    )
+    districts = serializers.PrimaryKeyRelatedField(
+        queryset=AdministrativeDistrict.objects.all(),
+        many=True,
+        allow_null=True,
+        required=False,
+    )
+    project_types = serializers.PrimaryKeyRelatedField(
+        queryset=ProjectType.objects.all(), many=True, allow_null=True, required=False
+    )
+    topics = serializers.PrimaryKeyRelatedField(
+        queryset=Topic.objects.all(), many=True, allow_null=True, required=False
+    )
+    status_display = serializers.CharField(source="get_status_display", read_only=True)
+
+    class Meta:
+        model = SearchProfile
+        fields = [
+            "id",
+            "user",
+            "name",
+            "description",
+            "disabled",
+            "status",
+            "status_display",
+            "query",
+            "query_text",
+            "organisations",
+            "districts",
+            "project_types",
+            "topics",
+        ]
+
+        read_only_fields = ["user"]
+
+    def create(self, validated_data):
+        # Pop many-to-many and one-to-many fields from validated_data
+        query_text = validated_data.pop("query_text", None)
+        organisations_data = validated_data.pop("organisations", None)
+        districts_data = validated_data.pop("districts", None)
+        project_types_data = validated_data.pop("project_types", None)
+        topics_data = validated_data.pop("topics", None)
+
+        if query_text:
+            query, _ = KiezradarQuery.objects.get_or_create(text=query_text)
+            validated_data["query"] = query
+
+        # need to create a profile before we can add m2m relations
+        search_profile = SearchProfile.objects.create(**validated_data)
+
+        if organisations_data:
+            search_profile.organisations.set(organisations_data)
+        if districts_data:
+            search_profile.districts.set(districts_data)
+        if project_types_data:
+            for project_type in project_types_data:
+                project_type, _ = ProjectType.objects.get_or_create(pk=project_type.id)
```

Comment:
the project types are predefined and created in the data migration, right ? Do we need `get_or_create` here? When I look at `project_types_data` it already contains `ProjectType` objects, so we don't need to fetch them from the db again, doing `set()` should work I think.

Reply:
true, i wrote this before the custom migration